src/Symfony/Component/Serializer/Normalizer/ChainDenormalizer.php

```diff
@@ -0,0 +1,207 @@
+        'bool' => true,
+        'float' => true,
+        'string' => true,
+    ];
+
+    /**
+     * @deprecated since Symfony 7.1
+     */
+    private ?SerializerInterface $serializer = null;
+
+    /**
+     * @var DenormalizerInterface[]
+     */
+    private array $denormalizers = [];
+
+    /**
+     * @var array<string, array<string, array<bool>>>
+     */
+    private array $denormalizerCache = [];
+
+    /**
+     * @var array<string, array<class-string|'*'|'object'|string, bool|null>>
+     */
+    private array $supportedCache = [];
+
+    /**
+     * @param DenormalizerInterface[] $denormalizers
+     */
+    public function __construct(array $denormalizers = [])
+    {
+        foreach ($denormalizers as $denormalizer) {
+            $this->addDenormalizer($denormalizer);
+        }
+    }
+
+    public function addDenormalizer(DenormalizerInterface $denormalizer): void
+    {
+        if ($denormalizer instanceof DenormalizerAwareInterface) {
+            $denormalizer->setDenormalizer($this);
+        }
+
+        if (null !== $this->serializer && $denormalizer instanceof SerializerAwareInterface) {
+            $denormalizer->setSerializer($this->serializer);
+        }
+
+        $this->denormalizers[] = $denormalizer;
+        $this->denormalizerCache = [];
+        $this->supportedCache = [];
+    }
+
+    /**
+     * @throws NotNormalizableValueException
+     */
+    public function denormalize(mixed $data, string $type, ?string $format = null, array $context = []): mixed
+    {
+        if (isset($context[DenormalizerInterface::COLLECT_DENORMALIZATION_ERRORS], $context['not_normalizable_value_exceptions'])) {
+            throw new LogicException('Passing a value for "not_normalizable_value_exceptions" context key is not allowed.');
+        }
+
+        $denormalizer = $this->getDenormalizer($data, $type, $format, $context);
+
+        // Check for a denormalizer first, e.g. the data is wrapped
+        if (!$denormalizer && isset(self::SCALAR_TYPES[$type])) {
+            if (!('is_'.$type)($data)) {
+                throw NotNormalizableValueException::createForUnexpectedDataType(sprintf('Data expected to be of type "%s" ("%s" given).', $type, get_debug_type($data)), $data, [$type], $context['deserialization_path'] ?? null, true);
+            }
+
+            return $data;
+        }
+
+        if ([] === $this->denormalizers) {
+            throw new LogicException('You must register at least one normalizer to be able to denormalize objects.');
+        }
+
+        if (!$denormalizer) {
+            throw new NotNormalizableValueException(sprintf('Could not denormalize object of type "%s", no supporting normalizer found.', $type));
+        }
+
+        return $denormalizer->denormalize($data, $type, $format, $context);
+    }
+
+    /**
+     * Returns a matching denormalizer.
+     *
+     * @param mixed       $data    Data to restore
+     * @param string      $class   The expected class to instantiate
```

Comment:
Related to my previous comment, this is not true if a scalar type is expected here.